# ML-09 — Validation and Research Claim Audit

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/133Haseeb/MLflyrankhaseeb/blob/main/work/notebooks/w06_validation_audit.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. Two paper findings + my methodology questions

*Pick two findings from the FlyRank research paper. For each: where does the label come from, and does the validation design carry the claim? Constructive tone.*

Finding 1

The paper shows that pages with declining search performance can be identified using machine learning models built from historical SEO data.

My methodology question

I would like to know how the target label was created. Was it based on real webpage refresh decisions or was it generated from historical performance trends? The answer helps explain how reliable the model predictions are.

Finding 2

The paper reports that machine learning models perform better than simple handwritten rules for identifying pages that need attention.

My methodology question

I would like to know whether the evaluation was performed using client-based or time-based validation. Testing on completely unseen clients provides a more realistic estimate of model performance.

## 2. My model under an honest split (before/after)

*Re-run your Week-5 model under a grouped or time-aware split. Show both numbers.*

In [2]:

%pip -q install duckdb huggingface_hub scikit-learn pandas

import duckdb
import os
from google.colab import userdata

HF_TOKEN = userdata.get("HF_Token")

con = duckdb.connect()

con.execute(f"""
CREATE OR REPLACE SECRET hf
(TYPE huggingface, TOKEN '{HF_TOKEN}')
""")

REL = "hf://datasets/FlyRank/internship-warehouse"

TABLES = {
    "fact_daily": f"read_parquet('{REL}/fact_content_daily_performance/**/*.parquet')",
    "dim_content": f"read_parquet('{REL}/dim_content.parquet')",
    "fact_query": f"read_parquet('{REL}/fact_content_query_90d.parquet')"
}

print("Connected Successfully!")

Connected Successfully!


In [3]:

features = con.sql(f"""

WITH latest_day AS (

SELECT
MAX(report_date) AS end_date
FROM {TABLES['fact_daily']}

),

daily_features AS (

SELECT

f.client_hash_id,
f.content_hash_id,

AVG(
CASE
WHEN report_date >
end_date - INTERVAL 20 DAY
THEN gsc_impressions
END
) AS avg_impressions,

AVG(
CASE
WHEN report_date >
end_date - INTERVAL 20 DAY
THEN gsc_clicks
END
) AS avg_clicks,

AVG(
CASE
WHEN report_date >
end_date - INTERVAL 20 DAY
THEN gsc_avg_position
END
) AS avg_position,

SUM(
CASE
WHEN report_date >
end_date - INTERVAL 20 DAY
THEN gsc_impressions
END
) AS imp_last20,

SUM(
CASE
WHEN report_date >
end_date - INTERVAL 40 DAY
AND report_date <= end_date - INTERVAL 20 DAY
THEN gsc_impressions
END
) AS imp_prev20,

SUM(
CASE
WHEN report_date >
end_date - INTERVAL 20 DAY
THEN gsc_clicks
END
) AS clicks_last20,

SUM(
CASE
WHEN report_date >
end_date - INTERVAL 20 DAY
THEN ga4_sessions
END
) AS sessions_last20,

SUM(
CASE
WHEN report_date >
end_date - INTERVAL 20 DAY
THEN ga4_pageviews
END
) AS pageviews_last20

FROM {TABLES['fact_daily']} f
CROSS JOIN latest_day

GROUP BY
client_hash_id,
content_hash_id

)

SELECT

d.*,

c.word_count,
c.char_count,
c.search_volume,
c.competition,
c.backlinks,
c.content_type,

q.content_visible_query_count,
q.rare_impressions_share,
q.anonymized_impressions_share

FROM daily_features d

LEFT JOIN {TABLES['dim_content']} c

ON
d.client_hash_id=c.client_hash_id
AND
d.content_hash_id=c.content_hash_id

LEFT JOIN (

SELECT

content_hash_id,

MAX(content_visible_query_count)
AS content_visible_query_count,

MAX(rare_impressions_share)
AS rare_impressions_share,

MAX(anonymized_impressions_share)
AS anonymized_impressions_share

FROM {TABLES['fact_query']}

GROUP BY content_hash_id

) q

ON d.content_hash_id=q.content_hash_id

""").df()

features.head()

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,client_hash_id,content_hash_id,avg_impressions,avg_clicks,avg_position,imp_last20,imp_prev20,clicks_last20,sessions_last20,pageviews_last20,word_count,char_count,search_volume,competition,backlinks,content_type,content_visible_query_count,rare_impressions_share,anonymized_impressions_share
0,client_08a6a72ff48e62c0,content_447b93d5ff670356,0.25,0.0,81.000000,5.0,49.0,0.0,NaN,NaN,<NA>,<NA>,10,0.15,<NA>,keyword article,2,0.162242,0.696165
1,client_08a6a72ff48e62c0,content_4483e354a2992497,7.75,0.0,9.707571,155.0,507.0,0.0,NaN,NaN,2261,14454,0,0.00,0,keyword article,13,0.051176,0.799447
2,client_08a6a72ff48e62c0,content_4489a53aa451dca0,0.95,0.0,10.583333,19.0,36.0,0.0,NaN,NaN,<NA>,<NA>,10,0.00,<NA>,keyword article,4,0.125000,0.625000
3,client_08a6a72ff48e62c0,content_44a8fa878cc63e32,0.05,0.0,4.000000,1.0,2.0,0.0,NaN,NaN,<NA>,<NA>,0,0.00,0,keyword article,2,0.097222,0.638889
4,client_08a6a72ff48e62c0,content_44afc7eca6a713b7,0.05,0.0,93.000000,1.0,24.0,0.0,NaN,NaN,<NA>,<NA>,20,0.19,<NA>,keyword article,1,0.589474,0.063158


In [4]:

import numpy as np

features["needs_refresh"] = np.where(

features["imp_last20"] <
0.80 * features["imp_prev20"],

1,

0

)

features["ctr"] = (

features["clicks_last20"] /

features["imp_last20"].replace(0,np.nan)

)

features.head()

,client_hash_id,content_hash_id,avg_impressions,avg_clicks,avg_position,imp_last20,imp_prev20,clicks_last20,sessions_last20,pageviews_last20,...,char_count,search_volume,competition,backlinks,content_type,content_visible_query_count,rare_impressions_share,anonymized_impressions_share,needs_refresh,ctr
0,client_08a6a72ff48e62c0,content_447b93d5ff670356,0.25,0.0,81.000000,5.0,49.0,0.0,NaN,NaN,...,<NA>,10,0.15,<NA>,keyword article,2,0.162242,0.696165,1,0.0
1,client_08a6a72ff48e62c0,content_4483e354a2992497,7.75,0.0,9.707571,155.0,507.0,0.0,NaN,NaN,...,14454,0,0.00,0,keyword article,13,0.051176,0.799447,1,0.0
2,client_08a6a72ff48e62c0,content_4489a53aa451dca0,0.95,0.0,10.583333,19.0,36.0,0.0,NaN,NaN,...,<NA>,10,0.00,<NA>,keyword article,4,0.125000,0.625000,1,0.0
3,client_08a6a72ff48e62c0,content_44a8fa878cc63e32,0.05,0.0,4.000000,1.0,2.0,0.0,NaN,NaN,...,<NA>,0,0.00,0,keyword article,2,0.097222,0.638889,1,0.0
4,client_08a6a72ff48e62c0,content_44afc7eca6a713b7,0.05,0.0,93.000000,1.0,24.0,0.0,NaN,NaN,...,<NA>,20,0.19,<NA>,keyword article,1,0.589474,0.063158,1,0.0


In [5]:

features.fillna(0, inplace=True)

selected_features = [

"avg_impressions",
"avg_clicks",
"avg_position",
"sessions_last20",
"pageviews_last20",
"ctr",
"word_count",
"char_count",
"search_volume",
"competition",
"backlinks",
"content_visible_query_count",
"rare_impressions_share",
"anonymized_impressions_share"

]

X = features[selected_features]

y = features["needs_refresh"]

In [6]:
from sklearn.model_selection import GroupShuffleSplit
from sklearn.ensemble import RandomForestClassifier

groups = features["client_hash_id"]

gss = GroupShuffleSplit(
    n_splits=1,
    test_size=0.20,
    random_state=42
)

train_idx, test_idx = next(
    gss.split(X, y, groups)
)

X_train = X.iloc[train_idx]
X_test = X.iloc[test_idx]

y_train = y.iloc[train_idx]
y_test = y.iloc[test_idx]

model = RandomForestClassifier(
    n_estimators=200,
    random_state=42,
    n_jobs=-1
)

model.fit(X_train, y_train)

pred = model.predict(X_test)

In [7]:
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix

print("Accuracy")
print(accuracy_score(y_test, pred))

print()

print("Classification Report")
print(classification_report(y_test, pred))

print()

print("Confusion Matrix")
print(confusion_matrix(y_test, pred))

Accuracy
0.8414378962491021

Classification Report
              precision    recall  f1-score   support

           0       0.89      0.93      0.91     55546
           1       0.36      0.25      0.30      8492

    accuracy                           0.84     64038
   macro avg       0.63      0.59      0.60     64038
weighted avg       0.82      0.84      0.83     64038


Confusion Matrix
[[51754  3792]
 [ 6362  2130]]


Honest Validation

In Assignment 5, I evaluated the model using a random train-test split and achieved an accuracy of 84%.

For this assignment, I used GroupShuffleSplit based on client_hash_id. This ensures that webpages from the same client do not appear in both the training and testing datasets.

This grouped validation provides a more realistic estimate of how well the model may perform on completely unseen clients. It reduces information sharing between the training and testing data and gives a more trustworthy evaluation of the model.

## 3. Leakage audit

*The same hunt from Week 3, on your final feature set.*

I reviewed the final feature set used in my warehouse model to ensure that no future information or target-related columns were included.

The model only uses information available before making the refresh decision, including:

Average impressions

Average clicks

Average position

Sessions

Pageviews

Word count

Character count

Search volume

Competition

Backlinks

Visible query count

Rare impression share

Anonymized impression share

The target variable needs_refresh was created by comparing the last 20 days of impressions with the previous 20 days. This target column was not used as an input feature during training.

Therefore, I did not find evidence of feature leakage in the final model.

## 4. Claim rewrite

*Take your own boldest sentence and rewrite it in safe language: observed, measured, directional, decision-support.*

Original Claim

My Random Forest model predicts which webpages should be refreshed.

Revised Claim

In this warehouse dataset, the Random Forest model learned useful patterns from aggregated SEO features. The predictions may help prioritize webpages for review, but they should be used as decision support rather than automatic decisions. Human review is still recommended before refreshing any webpage.

## Self-check

Before you submit, confirm each line honestly:

- [Checked ] Every section above is filled — markdown thinking AND the code that backs it
- [Checked ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [Checked ] No client names, URLs, or private queries anywhere
- [Checked ] My claims use careful words: observed, measured, directional, decision-support
- [Checked ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.